### PR TITLE
Optimize initial dashboard load - batch metadata calls, fix rendering issues with parameters

### LINKED ISSUE #41468: Performance hit after upgrade to 1.49.x
### Describe the bug

After upgrading from 1.48.4 to 1.49.5 our dashboards take much longer to load. One dashboard contains 19 questions. Before the upgrade it took a couple of seconds. Now it takes up to 20 seconds to fully load the page. It is not just my pc, got complains from many colleagues. It is also not just one dashboard, but multiple.

### To Reproduce

Make a heavier dashboard on version 1.48 and measure its performance. Then upgrade to 1.49 and compare.

Hard for me to make a comparison now, because it would require a downgrade to 1.48. Is that even possible?

Here at least a screenshot of the network calls.
image

I looked at one of the longer API requests of more than 5 seconds (see also the Metabase server logs). The underlying BigQuery query takes less than 1 second. It also did result in 0 rows, so the API response data json is small.

### Expected behavior

That the performance was not that impacted.

### Logs

[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/user/current 200 25.1 ms (11 DB calls) App DB connections: 1/15 Jetty threads: 4/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/session/properties 200 27.9 ms (8 DB calls) App DB connections: 1/15 Jetty threads: 4/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/collection/root 200 11.9 ms (2 DB calls) App DB connections: 3/15 Jetty threads: 6/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/bookmark 200 8.3 ms (1 DB calls) App DB connections: 1/15 Jetty threads: 5/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/database 200 8.3 ms (3 DB calls) App DB connections: 1/15 Jetty threads: 5/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/util/bug_report_details 200 3.9 ms (1 DB calls) App DB connections: 1/15 Jetty threads: 4/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/collection/tree 200 53.5 ms (6 DB calls) App DB connections: 3/15 Jetty threads: 7/50 (24 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/search 200 95.1 ms (5 DB calls) App DB connections: 2/15 Jetty threads: 7/50 (23 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/dashboard/76 200 261.4 ms (48 DB calls) App DB connections: 1/15 Jetty threads: 5/50 (23 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/pulse/form_input 200 3.1 ms (1 DB calls) App DB connections: 2/15 Jetty threads: 6/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:23+02:00 DEBUG metabase.server.middleware.log GET /api/collection/102 200 15.6 ms (4 DB calls) App DB connections: 2/15 Jetty threads: 6/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:24+02:00 DEBUG metabase.server.middleware.log GET /api/database/4/schemas 200 605.0 ms (4 DB calls) App DB connections: 14/15 Jetty threads: 27/50 (3 idle, 0 queued) (145 total active threads) Queries in flight: 2 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:24+02:00 DEBUG metabase.server.middleware.log GET /api/table/26967/query_metadata 200 810.1 ms (9 DB calls) App DB connections: 11/15 Jetty threads: 25/50 (3 idle, 0 queued) (145 total active threads) Queries in flight: 2 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:24+02:00 DEBUG metabase.server.middleware.log GET /api/table/26963/query_metadata 200 817.8 ms (11 DB calls) App DB connections: 10/15 Jetty threads: 25/50 (3 idle, 0 queued) (145 total active threads) Queries in flight: 2 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:24+02:00 DEBUG metabase.server.middleware.log GET /api/table/27089/query_metadata 200 911.4 ms (11 DB calls) App DB connections: 13/15 Jetty threads: 23/50 (5 idle, 0 queued) (145 total active threads) Queries in flight: 3 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:24+02:00 DEBUG metabase.server.middleware.log GET /api/table/26391/query_metadata 200 1.0 s (11 DB calls) App DB connections: 15/15 Jetty threads: 22/50 (6 idle, 0 queued) (145 total active threads) Queries in flight: 3 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:24+02:00 DEBUG metabase.server.middleware.log GET /api/field/366593 200 806.3 ms (7 DB calls) App DB connections: 13/15 Jetty threads: 21/50 (7 idle, 0 queued) (145 total active threads) Queries in flight: 5 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:24+02:00 DEBUG metabase.server.middleware.log GET /api/table/26828/query_metadata 200 998.2 ms (9 DB calls) App DB connections: 12/15 Jetty threads: 21/50 (7 idle, 0 queued) (145 total active threads) Queries in flight: 5 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/27104/query_metadata 200 1.1 s (9 DB calls) App DB connections: 15/15 Jetty threads: 20/50 (8 idle, 0 queued) (145 total active threads) Queries in flight: 4 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/field/366628 200 1.0 s (7 DB calls) App DB connections: 10/15 Jetty threads: 19/50 (9 idle, 0 queued) (145 total active threads) Queries in flight: 5 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26933/query_metadata 200 1.2 s (11 DB calls) App DB connections: 10/15 Jetty threads: 18/50 (10 idle, 0 queued) (145 total active threads) Queries in flight: 5 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/27087/query_metadata 200 1.0 s (11 DB calls) App DB connections: 9/15 Jetty threads: 18/50 (10 idle, 0 queued) (145 total active threads) Queries in flight: 5 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/27059/query_metadata 200 1.2 s (11 DB calls) App DB connections: 13/15 Jetty threads: 18/50 (10 idle, 0 queued) (145 total active threads) Queries in flight: 5 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26389/query_metadata 200 1.2 s (11 DB calls) App DB connections: 11/15 Jetty threads: 17/50 (11 idle, 0 queued) (145 total active threads) Queries in flight: 5 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26903/query_metadata 200 1.1 s (11 DB calls) App DB connections: 13/15 Jetty threads: 17/50 (11 idle, 0 queued) (145 total active threads) Queries in flight: 7 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/field/366581 200 1.2 s (7 DB calls) App DB connections: 15/15 Jetty threads: 16/50 (12 idle, 0 queued) (145 total active threads) Queries in flight: 7 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26982/query_metadata 200 1.4 s (11 DB calls) App DB connections: 15/15 Jetty threads: 16/50 (12 idle, 0 queued) (145 total active threads) Queries in flight: 7 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26968/query_metadata 200 1.4 s (11 DB calls) App DB connections: 10/15 Jetty threads: 12/50 (16 idle, 0 queued) (145 total active threads) Queries in flight: 6 (1 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/10165/query_metadata 200 1.4 s (11 DB calls) App DB connections: 10/15 Jetty threads: 15/50 (13 idle, 0 queued) (145 total active threads) Queries in flight: 7 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/9912/query_metadata 200 1.2 s (9 DB calls) App DB connections: 8/15 Jetty threads: 12/50 (16 idle, 0 queued) (145 total active threads) Queries in flight: 7 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/27094/query_metadata 200 1.2 s (11 DB calls) App DB connections: 15/15 Jetty threads: 14/50 (14 idle, 0 queued) (145 total active threads) Queries in flight: 7 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26936/query_metadata 200 1.4 s (11 DB calls) App DB connections: 8/15 Jetty threads: 9/50 (19 idle, 0 queued) (145 total active threads) Queries in flight: 8 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/27091/query_metadata 200 1.0 s (11 DB calls) App DB connections: 10/15 Jetty threads: 12/50 (16 idle, 0 queued) (145 total active threads) Queries in flight: 7 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/27058/query_metadata 200 1.4 s (9 DB calls) App DB connections: 14/15 Jetty threads: 13/50 (15 idle, 0 queued) (145 total active threads) Queries in flight: 7 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26979/query_metadata 200 1.5 s (9 DB calls) App DB connections: 14/15 Jetty threads: 11/50 (18 idle, 0 queued) (145 total active threads) Queries in flight: 8 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/card/1301 200 1.3 s (13 DB calls) App DB connections: 10/15 Jetty threads: 11/50 (18 idle, 0 queued) (145 total active threads) Queries in flight: 8 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/27115/query_metadata 200 1.5 s (9 DB calls) App DB connections: 9/15 Jetty threads: 11/50 (18 idle, 0 queued) (145 total active threads) Queries in flight: 8 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26934/query_metadata 200 1.1 s (11 DB calls) App DB connections: 6/15 Jetty threads: 6/50 (22 idle, 0 queued) (145 total active threads) Queries in flight: 9 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 INFO metabase.query-processor.middleware.cache Query took 424.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26981/query_metadata 200 1.7 s (11 DB calls) App DB connections: 12/15 Jetty threads: 8/50 (20 idle, 0 queued) (145 total active threads) Queries in flight: 8 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1298/card/1279/query 202 [ASYNC: completed] 1.8 s (21 DB calls) App DB connections: 10/15 Jetty threads: 4/50 (24 idle, 0 queued) (145 total active threads) Queries in flight: 8 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log GET /api/table/26829/query_metadata 200 1.8 s (11 DB calls) App DB connections: 5/15 Jetty threads: 5/50 (23 idle, 0 queued) (146 total active threads) Queries in flight: 9 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1275/card/1317/query 202 [ASYNC: completed] 1.9 s (20 DB calls) App DB connections: 7/15 Jetty threads: 4/50 (24 idle, 0 queued) (145 total active threads) Queries in flight: 7 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 INFO metabase.query-processor.middleware.cache Query took 437.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 INFO metabase.query-processor.middleware.cache Query took 290.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1304/card/1311/query 202 [ASYNC: completed] 2.1 s (22 DB calls) App DB connections: 5/15 Jetty threads: 3/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 9 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 INFO metabase.query-processor.middleware.cache Query took 300.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/357/card/372/query 202 [ASYNC: completed] 2.1 s (20 DB calls) App DB connections: 6/15 Jetty threads: 3/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 8 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:25+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/2658/card/2613/query 202 [ASYNC: completed] 2.2 s (20 DB calls) App DB connections: 6/15 Jetty threads: 3/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 7 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:26+02:00 INFO metabase.query-processor.middleware.cache Query took 412.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/361/card/374/query 202 [ASYNC: completed] 2.2 s (21 DB calls) App DB connections: 1/15 Jetty threads: 3/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 6 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 INFO metabase.query-processor.middleware.cache Query took 282.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:26+02:00 INFO metabase.query-processor.middleware.cache Query took 419.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 INFO metabase.query-processor.middleware.cache Query took 478.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 INFO metabase.query-processor.middleware.cache Query took 400.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/2028/card/1996/query 202 [ASYNC: completed] 2.5 s (22 DB calls) App DB connections: 5/15 Jetty threads: 5/50 (24 idle, 0 queued) (145 total active threads) Queries in flight: 6 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1307/card/1310/query 202 [ASYNC: completed] 2.5 s (21 DB calls) App DB connections: 5/15 Jetty threads: 5/50 (24 idle, 0 queued) (145 total active threads) Queries in flight: 5 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/791/card/830/query 202 [ASYNC: completed] 2.5 s (20 DB calls) App DB connections: 4/15 Jetty threads: 4/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 5 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:26+02:00 INFO metabase.query-processor.middleware.cache Query took 342.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 INFO metabase.query-processor.middleware.cache Query took 794.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log GET /api/dashboard/76 200 2.4 s (48 DB calls) App DB connections: 6/15 Jetty threads: 5/50 (24 idle, 0 queued) (145 total active threads) Queries in flight: 5 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1306/card/376/query 202 [ASYNC: completed] 2.6 s (20 DB calls) App DB connections: 4/15 Jetty threads: 5/50 (24 idle, 0 queued) (145 total active threads) Queries in flight: 4 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1358/card/1376/query 202 [ASYNC: completed] 2.6 s (25 DB calls) App DB connections: 5/15 Jetty threads: 5/50 (24 idle, 0 queued) (145 total active threads) Queries in flight: 3 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1302/card/1304/query 202 [ASYNC: completed] 2.6 s (21 DB calls) App DB connections: 4/15 Jetty threads: 4/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 4 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1328/card/1362/query 202 [ASYNC: completed] 2.7 s (22 DB calls) App DB connections: 4/15 Jetty threads: 4/50 (24 idle, 0 queued) (145 total active threads) Queries in flight: 2 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:26+02:00 INFO metabase.query-processor.middleware.cache Query took 604.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1305/card/1309/query 202 [ASYNC: completed] 2.8 s (21 DB calls) App DB connections: 3/15 Jetty threads: 4/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 3 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/2094/card/2072/query 202 [ASYNC: completed] 2.8 s (22 DB calls) App DB connections: 2/15 Jetty threads: 4/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 2 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 INFO metabase.query-processor.middleware.cache Query took 386.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:26+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1312/card/1308/query 202 [ASYNC: completed] 2.9 s (27 DB calls) App DB connections: 2/15 Jetty threads: 3/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 1 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:27+02:00 INFO metabase.query-processor.middleware.cache Query took 1.1 s to run; minimum for cache eligibility is 2.0 s; not eligible
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:27+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1311/card/1332/query 202 [ASYNC: completed] 3.2 s (21 DB calls) App DB connections: 2/15 Jetty threads: 3/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 1 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:27+02:00 INFO metabase.query-processor.middleware.cache Query took 726.0 ms to run; minimum for cache eligibility is 2.0 s; not eligible
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:27+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/pivot/76/dashcard/1310/card/1289/query 202 [ASYNC: completed] 4.0 s (26 DB calls) App DB connections: 2/15 Jetty threads: 3/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:28+02:00 DEBUG metabase.server.middleware.log GET /api/table/26388/query_metadata 200 53.8 ms (11 DB calls) App DB connections: 1/15 Jetty threads: 4/50 (25 idle, 0 queued) (145 total active threads) Queries in flight: 0 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:29+02:00 INFO metabase.query-processor.middleware.cache Query took 2.6 s to run; minimum for cache eligibility is 2.0 s; not eligible
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:29+02:00 DEBUG metabase.server.middleware.log POST /api/dashboard/76/dashcard/1313/card/1303/query 202 [ASYNC: completed] 5.2 s (26 DB calls) App DB connections: 2/15 Jetty threads: 4/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 0 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:29+02:00 DEBUG metabase.server.middleware.log GET /api/table/26966/query_metadata 200 31.3 ms (9 DB calls) App DB connections: 1/15 Jetty threads: 4/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 0 (0 queued)
[06106280-cca0-4acb-9608-a0fbbe7beb56] 2024-04-16T14:58:34+02:00 DEBUG metabase.server.middleware.log GET /api/action 200 33.7 ms (4 DB calls) App DB connections: 1/15 Jetty threads: 4/50 (25 idle, 0 queued) (143 total active threads) Queries in flight: 0 (0 queued)
[23ce4c3d-6373-4477-9aa8-e2f7e0702447] 2024-04-16T14:58:40+02:00 DEBUG metabase.server.middleware.log GET /api/table/26908/query_metadata 200 42.5 ms (11 DB calls) App DB connections: 1/15 Jetty threads: 4/50 (25 idle, 0 queued) (146 total active threads) Queries in flight: 0 (0 queued)

### Information about your Metabase installation

```JSON
- Browser: chrome Version 123.0.6312.124 (Official Build) (arm64)
- OS: MacOS 14.4.1 (23E224)
- Databases: Google BigQuery
- Metabase version: 1.49.5
- Metabase hosting: Docker
- Metabase internal database: Google CloudSQL PostgreSQL 13
```

### Severity

Medium

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ranquild**
Fixes https://github.com/metabase/metabase/issues/41468

Changes:
- Batch metadata parsing. MBQL lib is such a big bottleneck so making all `query_metadata` requests before showing charts leads to better performance than doing `query` requests in parallel. Also because of "question url" feature we cannot load metadata in background anymore.
- Move parameters outside of `Dashboard` component to not cause re-rendering when `metadata` changes

### CODE REVIEW

File: frontend/src/metabase-lib/v1/parameters/utils/targets.ts [hunk 1]
```diff
@@ -45,7 +45,6 @@ export function getParameterTargetField(
   }
 
   const fieldRef = target[1];
-  const query = question.query();
```
Comment (ranquild) [inline frontend/src/metabase-lib/v1/parameters/utils/targets.ts:48]: Avoid using MBQL lib for performance reasons

File: frontend/src/metabase/components/TokenField/TokenField.tsx [hunk 2]
```diff
@@ -360,7 +360,7 @@ class _TokenField extends Component<TokenFieldProps, TokenFieldState> {
   };
 
   onInputBlur = () => {
-    if (this.props.updateOnInputBlur && this.props.parseFreeformValue) {
```
Comment (ranquild) [inline frontend/src/metabase/components/TokenField/TokenField.tsx:363]: Existing bug that became apparent when the fields were fixed

File: frontend/src/metabase-lib/v1/parameters/utils/targets.ts [hunk 3]
```diff
@@ -60,8 +59,12 @@ export function getParameterTargetField(
 
   if (isConcreteFieldReference(fieldRef)) {
     const fieldId = fieldRef[1];
-    const tableId = Lib.sourceTableOrCardId(query);
-    return metadata.field(fieldId, tableId) ?? metadata.field(fieldId);
+    const resultMetadata = question.getResultMetadata();
+    const fieldMetadata = resultMetadata.find(field => field.id === fieldId);
```
Comment (ranquild) [inline frontend/src/metabase-lib/v1/parameters/utils/targets.ts:63]: This turned out to select fields more correctly than before.

File: frontend/src/metabase/dashboard/actions/data-fetching-typed.ts [hunk 4]
```diff
@@ -126,20 +123,6 @@ export const fetchDashboard = createAsyncThunk(
 
       fetchDashboardCancellation = null;
 
-      if (dashboardType === "normal" || dashboardType === "transient") {
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/data-fetching-typed.ts:129]: Do not load metadata upfront. Each dashcard loads metadata on demand.

File: frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx [hunk 5]
```diff
@@ -93,8 +96,13 @@ const DashCardMenu = ({
     [question, result, handleDownload],
   );
 
-  const menuItems = useMemo(
-    () => [
+  const menuItems = useMemo(() => {
+    // performance optimization - do not parse the query when the menu is closed
+    if (!isOpened) {
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx:101]: Avoid using MBQL lib for performance reasons

File: frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx [hunk 6]
```diff
@@ -170,7 +178,8 @@ DashCardMenu.shouldRender = ({
     !isPublic &&
     !isEditing &&
     !isXray &&
-    (canEditQuestion(question) || canDownloadResults(result))
+    (getHasDataAccess(question.metadata().databasesList()) ||
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx:181]: This is not 100% the same but it should be correct in 99% real world cases. In the worst case users would get an empty menu.

File: frontend/src/metabase/dashboard/selectors.ts [hunk 7]
```diff
@@ -425,17 +423,13 @@ export const getSelectedTabId = createSelector(
...
     if (dashboard && selectedTabId === null) {
-      return getInitialSelectedTabId(dashboard);
+      return dashboard.tabs?.[0]?.id || null;
     }
 
     return selectedTabId;
   },
 );
 
-export function getInitialSelectedTabId(dashboard: Dashboard | StoreDashboard) {
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/selectors.ts:435]: Not used - inlining

File: e2e/test/scenarios/dashboard-filters/dashboard-filters-nested.cy.spec.js [hunk 8]
```diff
@@ -78,8 +78,11 @@ describe("scenarios > dashboard > filters > nested questions", () => {
     // Add multiple values (metabase#18113)
     filterWidget().click();
     popover().within(() => {
-      cy.findByText("Gizmo").click();
-      cy.findByText("Gadget").click();
+      cy.findByPlaceholderText("Enter some text")
```
Comment (ranquild) [inline e2e/test/scenarios/dashboard-filters/dashboard-filters-nested.cy.spec.js:81]: Previously we incorrectly assumed it's Products.Category field but it's coming from a SQL query without metadata and we shouldn't match it.

**Review by uladzimirdev — APPROVED** [review]
LGTM

**Pull request merged**